### ISSUE TITLE
Issue with Publicise functionality of Jetpack with latest versions of PN

### ISSUE THREAD

**Issue by kilasuit** [ProjectNami/projectnami#319]
For further background please also see -https://github.com/Automattic/jetpack/issues/11044

I have tried deploying a new site on many occasions to test whether this was an issue with the deployment & whether this was limited to PN or not - I have also deployed another vanilla Wordpress instance as well where the functionality is actually working fine, which makes me think this is a PN issue not a Jetpack issue - however on trying to diagnose further I am unsure of specifics in where I should be looking to attempt further diagnosis.

Any suggestions would be helpful and if needs be I can spin up a test instance in Azure that can be used to further diagnose if needed.

**Comment by kilasuit** [comment]
I have just updated my issue on Jetpack side @patrickebates 

FWIW - I have implemented the suggested possible fixes in https://github.com/ProjectNami/projectnami/issues/237#issuecomment-321665844 with out any success as well as the further comment in that thread by @jherve and restarted the Website and still having this issue.

This is however becoming more time critical as I cannot look to publish the site with this issue so as an aside option I am looking at other similar functionalities that could be used instead of publicise but this would not be my preferred option going forward.

**Comment by patrickebates** [comment]
Regarding #237 that's what we added the Must-Use plugin into PN for, to set that value.
https://github.com/ProjectNami/projectnami/commit/8769db88e01b21a955e2733e5d8ea37aaa2a3f74

I have seen another site which had Publicise active for many years which recently had failures trying to publish new posts.  I think it was related to a change in how Publicise operates where it send the data to WP.com rather than posting directly to the social media systems, but we didn't dig any deeper.

**Comment by patrickebates** [comment]
Those entries simply mean that not everything in Jetpack is respecting the function that will call the correct API to write options.  Unfortunately Jetpack is notorious for direct DB access when it isn't necessary, and even against their own recommendations.

I disconnected an existing site from Jetpack, then reconnected it and reconfigured the Twitter connection.  Seems to be working, based on the fact that it wasn't at all until I performed the reconnect.  Kept getting what looked like XMLRPC errors.

**Comment by kilasuit** [comment]
It's plausible that it's this version of the plugin that isn't working - what version are you running on your site? On another site I run I have this working all ok though the version of Jetpack is quite old and won't update at all it seems from the UI

**Comment by jherve** [comment]
> I have just updated my issue on Jetpack side @patrickebates
> 
> FWIW - I have implemented the suggested possible fixes in [#237 (comment)](https://github.com/ProjectNami/projectnami/issues/237#issuecomment-321665844) with out any success as well as the further comment in that thread by @jherve and restarted the Website and still having this issue.
> 
> This is however becoming more time critical as I cannot look to publish the site with this issue so as an aside option I am looking at other similar functionalities that could be used instead of publicise but this would not be my preferred option going forward.

Hey there ! 
I just swing by this thread to say that you probably wanted to ping @jeherve and not me ;)

**Comment by kilasuit** [comment]
sorry for the incorrect ping @jherve - that'll teach me to type usernames in future

**Comment by patrickebates** [comment]
Turns out the site I tested on is running Jetpack 6.2.1
I will try a dev site with 6.9

Also, I've had lots of issues trying to update Jetpack within sites running on Azure Web Apps. The time required to update seems to take longer than the max execution time allowed there.  I always upload Jetpack via FTP.

**Comment by patrickebates** [comment]
Now I'm more confused.  Put 6.9 on a dev site, connected it to Wordpress, and now the Sharing section is missing on WP.com to even set up Sharing.  On the bright site, there's no error and no translation logs on the dev site, so everything appears to be operating locally as expected.

**Comment by kilasuit** [comment]
If 6.2.1 works then I may try to go and find that version of the plugin and drop that on the site instead.

This is opposed to the 4.1.1 version that I know works on an older blog of mine. 

On the dev site you've just put 6.9 on can you connect the site to your twitter account via the sharing page in the wp-admin sharing page once Jetpack is installed at `/wp-admin/options-general.php?page=sharing`  ??

**Comment by patrickebates** [comment]
No, and until you provided me with that URL I wasn't aware it might have been moved back to the local installation.  With 6.2.1, sharing was configured via the Wordpress.com portal which worked.  This might explain why the connected 6.9 site doesn't have Sharing as an option in that portal.

Even worse, no error logs were generated when the connection failed in the same way you described above.

**Comment by kilasuit** [comment]
So this is then something that they've changed in Jetpack and not something that you can see as being a PN issue?

**Comment by kilasuit** [comment]
I have FYI dropped back to the 6.2.1 version (I am deploying site completely from code) and pushing a new deployment out now to test with that version.

**Comment by patrickebates** [comment]
I'm not sure if it is a PN issue or not at this time.  I lean towards it being a PN issue, as that's how Jetpack usually goes, but without errors to look at it's going to be next to impossible to track it down.

As I reference above, and in other writings I have done, Wordpress.com (especially their VIP platform) require that plugins do not perform direct DB access and instead use the APIs.  But when it comes to their own premier plugins, Jetpack and Akismet, they don't follow their own guidelines.  Thankfully someone added the option into Jetpack to force using the Options API, which eliminated most of the PN issues with that plugin.

I can say that at this time the problem doesn't appear to be a database issue, but I am concerned that some function is being triggered which has a setting to suppress error messages.  Wouldn't be the first time I've seen that happen with callbacks which may or may not run through XMLRPC.

**Comment by kilasuit** [comment]
> You do not experience any issues when using Jetpack version 6.2.1. (fwiw I would recommend against using such an old version of the plugin, as you would be missing out on the updates, some of them security-related -- # #, that we've made to the plugin in the past year).

I deployed a new site with version 6.2.1 and still have this exact same issue so it seems not to be just version related however I have another site running a much older 4.1.1. version of Jetpack & older version of PN too where it works fine and is also hosted in Azure too.

Re 
> Right now, Jetpack relies on XML-RPC for almost all communications with WordPress.com. If XML-RPC is not fully accessible, you will experience issues.

I checked this and all seems to be fine via the troubleshooting that I had done previously,

Re 
> If so, we would need more details about the error to better understand what is happening here. The Jetpack plugin includes a "Debug" page that contains some information about your site and its connection to WordPress.com. You can find it by clicking on the Debug link at the bottom of the Jetpack > Dashboard page in wp-admin.

> Could you go there, copy the contents of the advanced debug results at the bottom of the page, and send them to us via this contact form so we can dig into it some more. You can mention this Github issue in your email so my colleagues will have more background about the problem.

> @kilasuit I know you'd rather continue this conversation here on GitHub, as you mentioned here, and I'll be happy to come back to continue this conversation here once I know more, but for now we would need to have more information about the site to do some more digging, and that information is not something you can share publicly as it contains private details about the site's connection to WordPress.com.

I will send through the details via the contact form and mention this issue in there further updates

**Comment by kilasuit** [comment]
@jeherve  - I have sent this through now 👍

**Comment by jeherve** [comment]
> I have another site running a much older 4.1.1. version of Jetpack & older version of PN too where it works fine

What happens if you update Jetpack on that site? Does the problem start happening? If things are still good after you update Jetpack, the next step would be to try updating PN and see what happens. This would allow us to narrow down the issue a bit.

**Comment by kilasuit** [comment]
I am not confident on that site as it currently stands as the UI does not show that Jetpack can be updated at all, plus this is my list of things that I need to look at updating anyway. this is something that I hope to get some chance to look into in the next week or so as I move the configuration of the site more into a Config and deploy from code way than it is currently set up.

**Comment by kilasuit** [comment]
I've started with the 1.98 version of PN (so before WP v5) and have just kicked off a deployment of this. Will update if that works or not

**Comment by kilasuit** [comment]
Still getting the same issue - I'm gonna look at logs and see if I can find any reason why this may be happening

**Comment by patrickebates** [comment]
Even with WP_DEBUG set I'm not seeing any logs generated on my dev system.

I did make note of the URL which is reporting the error "There was a problem creating your connection. Please go back and try again or contact support."  Without exposing every parameter, it appears to be on WP.com

https://public-api.wordpress.com/connect/?magic=keyring&action=verify&service=twitter

**Comment by kilasuit** [comment]
Yes this seems to be the case for me too which makes sense as the Publicise feature links to WP.com account via Jetpack anyway. I also notice that this was not just affecting Twitter accounts but also Facebook too.

FWIW - I am using only the Free version of Jetpack (which says it supports this functionality)

**Comment by patrickebates** [comment]
See the third post here which references characters not allowed in passwords over ODBC.

https://social.msdn.microsoft.com/Forums/sqlserver/en-US/6a26aa9d-2640-4f1c-8a9a-5637c412266f/sql-passwords?forum=sqlsecurity

**Comment by kilasuit** [comment]
@jeherve - I cloned the site and updated Jetpack from 4.1.1 to 6.9 then deactivated and reactivated Jetpack, connected it to WP.com and then tried to set up sharing. it's not possible but does give a new error

```
Something which should never happen, happened. Sorry about that. If you try again, maybe it will work.

Error code: -10520
```

**Comment by jeherve** [comment]
> I did make note of the URL which is reporting the error "There was a problem creating your connection. Please go back and try again or contact support."
> makes sense as the Publicise feature links to WP.com account via Jetpack anyway. I also notice that this was not just affecting Twitter accounts but also Facebook too.

That's all correct. And if your site is not correctly connected to WordPress.com, it would indeed impact all Publicize connections, not just one.

> it's not possible but does give a new error

This error indicates connection issues between the site and WordPress.com. Could you now try to disconnect Jetpack from WordPress.com? To do so, click on Jetpack from your site's Dashboard and scroll all the way down where you'll see a link that says "Disconnect from WordPress.com". Click that to disconnect.

To reconnect, click the green button that says "Set up Jetpack" and follow the prompts. Are you able to reconnect?

**Comment by kilasuit** [comment]
@jeherve - I just did an update to my live site and this has borked the Jetpack install completely. I had to manually remove the Jetpack folder via FTP connection and reinstall the plugin from scratch

**Comment by jeherve** [comment]
It sounds like the plugin update process didn't complete properly, and some files were missing from your Jetpack install? This can happen sometimes, as discussed here:
https://wordpress.org/support/topic/fatal-error-updating-from-jetpack-6-8-1-to-jetpack-6-9/

**Comment by kilasuit** [comment]
@jeherve - the Jetpack update seems to have worked all ok on my original site where I've done both a PN update and and Jetpack update. 

This now seems to limit it to new sites however in the upgrade it failed to properly connect back up initially however WP.com shows that the site is still running on the older version of PN

That aside I still cannot get any new PN site to connect to Jetpack and be able to us Publicise at all and I need this for a new blog site that needs to go live soon. I may have to go live with other similar functionality but this is sub optimal at best especially as I will need a mechanism so similar to Publicise to allow authors to add their own social media options for items that they publish as well as the site's account.

**Comment by jeherve** [comment]
Did you have the chance to email me a detailed report from a site that failed, so I can take a closer look at what is wrong with the connection?

**Comment by kilasuit** [comment]
I sent it in via that support form and mentioned this issue so you should have it.

**Comment by patrickebates** [comment]
I did the same, providing the debug data from one of our dev sites.

**Comment by kilasuit** [comment]
Also whilst I have been able to connect my older site back up to Twitter and LinkedIn - Publicise has failed to publicise updates via either channel 🤷‍♂️

**Comment by kilasuit** [comment]
@jeherve - can do you have any updates at all?

**Comment by kilasuit** [comment]
@jeherve previous ticket was 1614272 though not heard from most recent submission

**Comment by patrickebates** [comment]
I also did not receive anything back from my recent submission

**Comment by kilasuit** [comment]
I however on reread of all emails got pointed to try hitting this API endpoint once Jetpack was installed on the site which results in a 403 on all of the test sites I have - https://public-api.wordpress.com/rest/v1/sites/<sitename/siteurl>

so for example I run in powershell

```powershell
invoke-webrequest https://public-api.wordpress.com/rest/v1/sites/mhasl.me
```

which returns a 403 error

which is pushing me to think that this is something either with Azure Websites or PN

**Comment by jeherve** [comment]
> previous ticket was 1614272 though not heard from most recent submission

I can't find any more recent email from you in our system. Could you try to email again?

Thank you.

**Comment by patrickebates** [comment]
Seeing this message at the Contact form link you supplied earlier in this thread...

`Support for customers on Jetpack's Free Plan is temporarily closed.
`

**Comment by jeherve** [comment]
Yeah, you'd need a paid plan right now. If you don't have a paid plan, could you email the email address you received our previous replies from?

**Comment by patrickebates** [comment]
@jeherve Please send a DM to ProjectNami on Twitter.  Should be able to securely provide email address there.

**Comment by jeherve** [comment]
I just messaged you there 👍

**Comment by kilasuit** [comment]
FYI latest Jetpack v7 still has this issue too

**Comment by kilasuit** [comment]
@jeherve I have just installed PN on a full version of Windows Server 2016, turned off all firewall settings and installed JP v7.1 

This seems to be able to enable the publicise feature which narrows this down to an Azure WebApp issue I think - I will dig more and test JP 7.1 on my development instance and see what happens

**Comment by kilasuit** [comment]
interestingly enough the sharing option in Wordpress.com disappears and reappears based on certain navigation pages that I click though but i still end up getting this issue when I attempt to connect Publicise with Twitter

![image](https://user-images.githubusercontent.com/6355225/52889935-2a7b3700-317a-11e9-8d9d-25a0e7d830d1.png)

**Comment by kilasuit** [comment]
The difference between working and non-working sites is as below

![image](https://user-images.githubusercontent.com/6355225/52890034-8ba30a80-317a-11e9-8c4d-1873ed8dff2b.png)


this says to me theres something wrong with Jetpack and how it's connecting to the site after that API call is carried out which isn't something that I can diagnose further I don't think

**Comment by patrickebates** [comment]
Ryan, I've been able to duplicate your results across a few different sites that I deal with on a routine basis.

Is there anything that these sites that are working vs not working have in common?  For example, are the sites which are working originally set up by one method, and the other sites set up a different way?  Perhaps some created via the Azure portal and others using our Deploy To Azure or another ARM template?

**Comment by kilasuit** [comment]
So my original blog site where Jetpack & Publicise was working fine seems to be not posting any new blogs now via Publicise at all after updating to 6.9 I then disconnected Jetpack, reconnected, reconnected all the Publicise connections (which seemed to work ok) but then subsequent posts don't seem to be triggering the Publicise functionality 🤷‍♂️ 

Mhasl.me (and minor variations) are all triggered from a slightly modified version of the `azuredeploy.json` in this repo that does things like get a secret from Key Vault for the db password and switching out some other minor details like site name etc for my dev / test / prod environments. This is all controlled and deployed via Azure DevOps Pipelines.

The only difference in all sites that I've trialled with seems to be PN v vanilla WP - though next on my trial list is WP on Windows Server with MySql (though that will be a tomorrow job)

Deploying a Wordpress Web App (as per the offering the Azure Portal) and then adding Jetpack to that site works, which narrows this down to PN or perhaps a windows issue

**Comment by patrickebates** [comment]
Let me complicate your testing a little bit...

Vanilla WP site with MySQL, deployed manually - working
PN site deployed via ARM - not working
PN site deployed manually - working

To be perfectly clear, Working is defined at this point as Publicise will connect and the IWR command you listed above works.

On the PN site that isn't working with the IWR command, I can't find anything in the logs at all.  I have even enabled full logging features in Azure Web Apps, including failed request tracking, and I can't find anything reporting the 403 errors we are seeing.

Going to manually deploy some more sites over the weekend to try and duplicate these results.

**Comment by kilasuit** [comment]
Here's my results 
>Vanilla WP site with MySQL, deployed manually to App Service

Publicise connects and `iwr` command working 

>PN site deployed via ARM to App Service - PHP set to 7.2

Publicise cannot connect and `iwr` is  not working
>PN site deployed manually to Window Server 2016 box with IIS & PHP 7.2

Publicise cannot connect and `iwr`  Not working

>PN site that was deployed a long time ago via Azure Deploy to Azure WebApp and recently upgraded to latest PN and JP versions and is on PHP 7.2
 Publicise is connected but not working anymore and `iwr` command Working - confusingly

Now the only difference I could see in the configuration was that on my older site I have this additional plugin
![image](https://user-images.githubusercontent.com/6355225/52906989-e35f7580-324f-11e9-8c8c-b3897953d3cb.png)

Now I thought that this this functionality was not only included with WP core but also within Jetpack with the Json API feature but I am now doubting this - @jeherve can you comment on this ??

**Comment by patrickebates** [comment]
I have finally managed to capture an error that might be of interest.  401.3 - Unauthorized by what might be a loopback request to rest_route=/jetpack/v4/connection

You do not have permission to view this directory or page because of the access control list (ACL) configuration or encryption settings for this resource on the Web server.

**Comment by patrickebates** [comment]
To further complicate things, it appears that the site which the IWR command works on is not able to view or update Settings on Wordpress.com.  But the site which the IWR command doesn't work with is able to view and update Settings via Wordpress.com.  Starting to think that isn't worthwhile in this effort.

Interestingly, both sites appear to be attempting full syncs with WP.com but neither is progressing.  Shouldn't be an issue as each site has at most a few dozen posts & pages.

**Comment by kilasuit** [comment]
I get the same issue with the full sync's too that neither complete, perhaps this is due to the same underlying error, I don't expect that it would be something else on top of that too or unlikely to be.

I'm unsure as where next to look into this with as it very much feels like an OS level issue. Have you managed to install PN on a Linux OS at all?

**Comment by patrickebates** [comment]
Yes, though the last time was a couple of years ago when we were involved in trials with the SQL Server team during their NDA period for their Linux release.  Worked like a champ once we also worked with their driver team to verify that SQLSRV for PHP was ready for Linux as well.  Though we didn't bother with much plugin testing at that time.

To burst your bubble a bit regarding this being an OS level issue, I did deploy vanilla WP with Azure DB for MySQL on a Windows-based Azure Web App and successfully connected social media accounts.  Didn't try to publish a post, though.

What's annoying me at the moment is that there is absolutely **_NO PHP ERROR LOGGING_** taking place with regards to these failures, even though I have WP_DEBUG set.  I have an email from @jeherve asking me to add some logging into the Jetpack plugin itself, so that looks like my next move.

At this point, I think all we need is some error output from exactly where the failure is taking place and we can resolve these issues.  Just wish I knew why we aren't seeing it.

**Comment by patrickebates** [comment]
I might have finally found something useful.  
class.jetpack.php
function add_nonce isn't respecting JETPACK_DISABLE_RAW_OPTIONS
This appears to be the source of the error logs you referenced at the start of this thread.

**Comment by patrickebates** [comment]
I edited the DB calls to confirm they would write the nonce correctly, and it appears they might have been working already.  I can see a nonce written with each attempt to connect to Twitter, but the connection still fails.

Still no error logs.

**Comment by jeherve** [comment]
> I thought that this this functionality was not only included with WP core but also within Jetpack with the Json API feature but I am now doubting this - @jeherve can you comment on this ??

Those are different APIs. That plugin you're using is indeed not necessary anymore; the API and all the endpoints included in that plugin are now part of WordPress core, since version 4.4.
The API that ships with Jetpack is a different one; it is made to interact with WordPress.com.

> class.jetpack.php
> function add_nonce isn't respecting JETPACK_DISABLE_RAW_OPTIONS

That could be what's causing issues.

> I can see a nonce written with each attempt to connect to Twitter, but the connection still fails.

If you look at the access logs for your site when trying to make that connection, what kind of response do you get?

**Comment by kilasuit** [comment]
As per the merge above into Jetpack as referenced by @jeherve I will look to integrate the latest version of Jetpack into my site deployment pipeline and give this a test out and see if this completes over the weekend.

**Comment by kilasuit** [comment]
This no longer affects me as I have migrated from my blogs using Project Nami entirely. Not sure if you want to keep this open or close it @patrickebates
